Display build version of workers in WebUI and capacity command

kaijchen opened the pull request:
### What changes are proposed in this pull request?

1. Add version and revision information in workerRegister proto and API. 
2. Add worker version and revision in `alluxio fsadmin report capacity`.
3. Add worker version and revision in workers tab of WebUI.

Fix #13763.

### Why are the changes needed?

1. When performing rolling upgrades, we need to know the status of the cluster.
(e.g. Which worker is still running the old version? Is recent upgrade sucessful?)
2. It's cumbersome to query each worker individually.
(And the WebUI of workers may not be exposed to the client due to firewall policy)
3. The revision is for distinguishing different builds of the same SNAPSHOT version.
(Useful for development and testing)

### Does this PR introduce any user facing changes?

1. Add worker version and revision in `alluxio fsadmin report capacity`.
```console
$ ./bin/alluxio fsadmin report capacity
Capacity information for all workers: 
    Total Capacity: 20.74GB
        Tier: MEM  Size: 20.74GB
    Used Capacity: 1804B
        Tier: MEM  Size: 1804B
    Used Percentage: 0%
    Free Percentage: 100%

Worker Name      Last Heartbeat   Storage       MEM              Version          Revision                                
9.134.131.52     0                capacity      20.74GB          2.9.0-SNAPSHOT   ecafed558c467208da86d23829c589d22634561c
                                  used          1804B (0%)
```
2. Add worker version and revision in workers tab of WebUI.
<img width="1451" alt="image" src="https://user-images.githubusercontent.com/5821159/201295810-c459ece5-5183-45a1-b9fd-75c8dc26fa72.png">

kaijchen commented:
UI change example:

<img width="1459" alt="image" src="https://user-images.githubusercontent.com/5821159/200992170-73388493-5a67-4d0c-980b-a3f642a05810.png">


```console
$ ./bin/alluxio fsadmin report workers
Alluxio workers: 
    WorkerInfo{id=1070416111312515592, address=WorkerNetAddress{host=9.134.131.52, containerHost=, rpcPort=29999, dataPort=29999, webPort=30000, domainSocketPath=, tieredIdentity=TieredIdentity(node=9.134.131.52, rack=null)}, lastContactSec=0, state=In Service, capacityBytes=22264296789, usedBytes=1804, startTimeMs=1668048392202, capacityBytesOnTiers={MEM=22264296789}, usedBytesOnTiers={MEM=1804}, version=2.9.0-SNAPSHOT, revision=d44b43ffd5d38b414cfee5360610ad2f4b1f0829}
```

kaijchen commented:
Cc @jiacheliu3 @apc999 PTAL.

maobaolong commented:
@kaijchen Would you please add more description to show why we need to display the revision and worker version

kaijchen commented:
> @kaijchen Would you please add more description to show why we need to display the revision and worker version

Thanks for the suggestion, I have updated the description. Feel free to ask more questions if it's not clear enough.

kaijchen commented:
I have updated this PR based on the discussion with @jiacheliu3 in community meeting. The changes are:

1. Include BuildVersion in Options, and avoid the API change.
2. Include version info in `alluxio fsadmin report capacity`, and avoid adding a new command.

PR discription has been updated.

maobaolong reviewed (approved):
This feature looks great, we can get the version of workers through this command tool and web ui. Thank you for your contribution!

tieujason330 reviewed (approved):
ui changes LGTM, thanks

dbw9580 commented on core/server/master/src/main/java/alluxio/master/block/meta/MasterWorkerInfo.java:
```diff
@@ -164,6 +168,7 @@ public MasterWorkerInfo(long id, WorkerNetAddress address) {
     mBlocks = new LongOpenHashSet();
     mToRemoveBlocks = new LongOpenHashSet();
     mLastUpdatedTimeMs = new AtomicLong(CommonUtils.getCurrentMs());
+    mBuildVersion = new AtomicReference<>(BuildVersion.getDefaultInstance());
```
add this to the `toString` output.

dbw9580 reviewed (changes requested):
Mostly LGTM with a few minor comments, thanks!

kaijchen commented:
Thanks @dbw9580 for the suggestion, I have addressed those issues.
PS: The javadoc of revision is originally copied from
https://github.com/Alluxio/alluxio/blob/a95ed9a15bb5c96c85342748214115b6e41e4ad0/core/common/src/main/java-templates/alluxio/ProjectConstants.java#L20-L21

jiacheliu3 reviewed (commented):
One last comment from me is on back/forward compatibility. By using `optional` gRPC field and `optins.hasBuildVersion()` I think old/new master can work with old/new workers. But does old/new CapacityCommand work with old/new worker?

All other changes LGTM and we can merge once we confirm on the above question. Thanks a lot for the good work!

kaijchen commented:
Thanks @jiacheliu3 for your questions.
I have tested v2.9.0 as old master/client with the current branch (e9e154f):

## New Master

### New Client

```console
$ bin/alluxio fsadmin report capacity
Capacity information for all workers: 
    Total Capacity: 20.74GB
        Tier: MEM  Size: 20.74GB
    Used Capacity: 1804B
        Tier: MEM  Size: 1804B
    Used Percentage: 0%
    Free Percentage: 100%

Worker Name      Last Heartbeat   Storage       MEM              Version          Revision                                
9.134.131.52     0                capacity      20.74GB          2.9.0-SNAPSHOT   8ae79c38ef6340a98665f39a96a305cbc1042bc7
                                  used          1804B (0%)
```

### Old Client

```console
$ bin/alluxio fsadmin report capacity
Capacity information for all workers: 
    Total Capacity: 20.74GB
        Tier: MEM  Size: 20.74GB
    Used Capacity: 1804B
        Tier: MEM  Size: 1804B
    Used Percentage: 0%
    Free Percentage: 100%

Worker Name      Last Heartbeat   Storage       MEM
9.134.131.52     0                capacity      20.74GB
                                  used          1804B (0%)
```

## Old Master

### New Client

```console
$ bin/alluxio fsadmin report capacity
Capacity information for all workers: 
    Total Capacity: 20.74GB
        Tier: MEM  Size: 20.74GB
    Used Capacity: 0B
        Tier: MEM  Size: 0B
    Used Percentage: 0%
    Free Percentage: 100%

Worker Name      Last Heartbeat   Storage       MEM              Version          Revision                                
9.134.131.52     0                capacity      20.74GB                                                                   
                                  used          0B (0%) 
```

### Old Client

```console
$ bin/alluxio fsadmin report capacity
bin/alluxio fsadmin report capacity
Capacity information for all workers: 
    Total Capacity: 20.74GB
        Tier: MEM  Size: 20.74GB
    Used Capacity: 0B
        Tier: MEM  Size: 0B
    Used Percentage: 0%
    Free Percentage: 100%

Worker Name      Last Heartbeat   Storage       MEM
9.134.131.52     0                capacity      20.74GB
                                  used          0B (0%)
```

kaijchen commented:
We **can** check `hasBuildVersion` in `GrpcUtils#fromProto()`, however, there isn't any `hasField` check for other optional fields either.

jiacheliu3 reviewed (approved):
LGTM, thanks!

jiacheliu3 commented:
alluxio-bot, merge this please

kaijchen commented:
> We **can** check `hasBuildVersion` in `GrpcUtils#fromProto()`, however, there isn't any `hasField` check for other optional fields either.

@jiacheliu3 The new default value is correct is because of we are using protobuf3 to compile our proto2 file. If using protobuf 2 to compile it, there will be problem.
